Core/Scripts: Survival Hunter Flanking Strike and related spells

MartGon opened the pull request:
<!--- (**********************************)
      (** Fill in the following fields **)
      (**********************************) --->

**Changes proposed:**

-  Animal Instincts http://www.wowhead.com/spell=204315/animal-instincts
-  Aspect of the Beast http://www.wowhead.com/spell=191384/aspect-of-the-beast
- Aspect of the Eagle http://www.wowhead.com/spell=186289/aspect-of-the-eagle
-  Flanking strike http://www.wowhead.com/spell=202800/flanking-strike
-  Flanking strike pet http://www.wowhead.com/spell=204740/flanking-strike
-  Mongoose bite http://www.wowhead.com/spell=190928/mongoose-bite
-  Hunting companion gain charge http://www.wowhead.com/spell=191335/hunting-companion
-  Hunting companion pet aura  http://www.wowhead.com/spell=191336/hunting-companion
-  Added SpellHistory::ModifyCharge() method for charge times reductions.
-  Added Flanking strike pet's aura to pet http://www.wowhead.com/spell=240152/flanking-strike, checked on client to allow Flankig Strike (202800) cast.

**Target branch(es):** 3.3.5/master

- [ ] 3.3.5
- [x] master

**Issues addressed:** Closes 
- #18867 It isn't exactly a fix for this issue, but the spell no longer makes the server crash


**Tests performed:** Tested in-game.


**Known issues and TODO list:** 

- [ ] Need Feedback about how to add the pet aura correctly
- [ ] Need Feedback about the extra threat on Flanking Strike pet
- [ ] Player's versatility it's always 0 when using that function


<!--- Notes
- Enable the setting "[√] Allow edits from maintainers." when creating your pull request.
- If this PR only contains SQL files, open a new issue instead and post or link the SQL in the issue.
- When adding new SQL files, name them 9999_99_99_99_db_name.sql to reduce the chance of possible merge conflicts.
--->

Commit:
Scripted some survival hunter spells

Commit:
Fixed some coding style mistakes

Commit:
More coding style fixes

Aokromes commented on sql/updates/world/master/2017_99_99_99_world.sql:
```diff
@@ -0,0 +1,29 @@
+--Hunter Spell Flanking Strike (202800)
```
1 delete+insert plz.

MartGon replied:
Do you mean something like:
```
INSERT INTO `spell_script_names` (`spell_id`, `ScriptName`) VALUES
('202800','spell_hun_flanking_strike');
('204740','spell_hun_flanking_strike_pet');
```
and
```
DELETE FROM `spell_script_names` WHERE `ScriptName`='spell_hun_flanking_strike' OR
ScriptName`='spell_hun_flanking_strike_pet';
```

Aokromes replied:
check https://github.com/TrinityCore/TrinityCore/pull/19332/files

Commit:
SQL and ModifyCharge mistake

Commit:
Another quick fix

Commit:
Update 2017_99_99_99_world.sql

- insert missing space between comment tag and text
- remove apostrophes ' ' from numbers (TC codestyle)
- shorten comment text (remove "Hunter spell ")
- add separator comment line on top

tkrokli commented:
I forgot to comment the actual  SQL error in the original file;
`- comma instead of semicolon between the inserted rows`

Shauren commented on src/server/scripts/Spells/spell_hunter.cpp:
```diff
@@ -270,6 +326,282 @@ class spell_hun_disengage : public SpellScriptLoader
+            {
+                return ValidateSpellInfo
+                ({
+                    SPELL_HUNTER_ASPECT_BEAST        ||
```
this isnt doing what you think its doing

MartGon replied:
Yes, my bad. It should be separated by commas instead of "||",right?

codeman8214 replied:
Yes in this case I would use commas

Shauren replied:
yes

joschiwald commented on src/server/scripts/Spells/spell_hunter.cpp:
```diff
@@ -270,6 +326,282 @@ class spell_hun_disengage : public SpellScriptLoader
+                            int32 ap = owner->ToPlayer()->GetInt32Value(UNIT_FIELD_ATTACK_POWER);
+                            float vers = owner->ToPlayer()->GetFloatValue(PLAYER_VERSATILITY);
+                            int32 level = caster->getLevel();
+                            //formula taken from wowhead
```
don't trust wowhead, use the formula from tooltip

MartGon replied:
Not the best comment, formula taken from spellwork is the same:

> and your pet deals $damage Physical damage. 
$lowNerf=${$min($pl,20)*0.05}
$damage=${1.5*(($AP*2.152))*(1+$@versadmg)*$lowNerf}

codeman8214 replied:
Yes that is correct the formula is in the right format for both

Commit:
Fixed validate calls. Changed comment

Commit:
Merge branch 'Survival_Hunter' of https://github.com/MartGon/TrinityCore into Survival_Hunter

Commit:
Corrected AddSc order

joschiwald commented on src/server/scripts/Spells/spell_hunter.cpp:
```diff
@@ -270,6 +326,282 @@ class spell_hun_disengage : public SpellScriptLoader
+                            float vers = owner->ToPlayer()->GetFloatValue(PLAYER_VERSATILITY);
+                            int32 level = caster->getLevel();
+                            //formula taken from tooltip
+                            float result = 1.5*(ap*2.152)*(1 + vers)*(std::min(20, level)*0.05f);
```
i have a different formula
```
$bmMastery=$?s76657[${1+0.01*$76657m1}][${1}]
$lowNerf=${0.5+$min($pl,20)*0.025}
$damage=${1.5*($83381m1+($RAP*3))*$<bmMastery>*$<lowNerf>*(1+$@versadmg)}
```

MartGon replied:
I think that formula is from kill command. The one I used is taken from Flanking Strike(202800) and it's applied to Flanking Strike (204740) which originally deals just 1 damage.

soulfrost replied:
float vers = owner->ToPlayer()->GetFloatValue(PLAYER_VERSATILITY); bad idea... versality should be implented in generic data calcs

MartGon replied:
Do you mean that versatility shouldn't be part of the formula because it's applied on every spell that deals damage? If that's true it's not longer needed on the formula.

MartGon replied:
@soulfrost Well, after thinking about it again... versatility may be a component of the formula because that spell is casted by the pet, so I don't know exactly if owner's versatility may affect it's dealed damage.

soulfrost replied:
versa affects all dealed/taken dmg, heal, absorm and affected on pet dmg too

MartGon replied:
@soulfrost Ok, I didn't know that versatily worked that way. Then I'm going to commit the new formula without that component. Thanks for making it clear!

Commit:
Removed versatility component from flanking strike pet damage formula

soulfrost commented:
@infernus55 versa should be deleted not from this PR - should be rewritten in other PR & removed from all other scripts

MartGon commented:
Closing this until further testing.
